docs/actions-update.md

```diff
@@ -0,0 +1,75 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+-->
+## Updating Action Language Runtimes
+
+OpenWhisk supports [several languages and runtimes](actions.md#languages-and-runtimes) that can be made
+available for usage in an OpenWhisk deployment. This is done via the [runtimes manifest](actions-new.md#the-runtimes-manifest).
+
+Over time, you may have the need for change:
+
+* Update runtimes to address security issues - for example, the latest code level of Node.js 10.
+* Remove runtime versions that are no more supported - for example, Node.js 6.
+* Add more languages due to user demand.
+* Remove languages that are no more needed.
+
+While adding and updating languages and runtimes is pretty straightforward, removing or renaming languages and runtimes
+requires some caution to prevent problems with existing actions.
+
+### Updating runtimes
+
+Follow these steps to update a particular runtime kind:
+
+1. Update the runtimes' container image.
+2. Update the corresponding `image` property in the [runtimes manifest](actions-new.md#the-runtimes-manifest) to use the new container image.
+3. Restart / re-deploy controllers and invokers such that they pick up the changed runtimes manifest.
+
+Already existing actions of the changed runtime kind will immediately use the new container image when the invoker creates a new action container.
+
+Apparently, this approach should only be used if existing actions do not break with the new container image. If a new container image is supposed to break existing actions, introduce a new runtime kind.
+
+### Removing runtimes
+
+Follow these steps to remove a particular runtime kind under the assumption that actions with the runtime kind exist in the system. Apparently, it makes sense to wait some time between accomplishing these steps to give action owners time to react.
+
+1. Deprecate the runtime kind by setting `"deprecated": true` in the [runtimes manifest](actions-new.md#the-runtimes-manifest). This setting prevents that new actions can be created with the deprecated action kind. In addition, existing actions cannot be changed to the deprecated action kind any more.
+2. Ask owners of existing actions with runtime kind to be removed to update their actions to a different action kind.
+3. Create an automated process that identifies all actions with the runtime kind to be removed in the system's action artifact store. Either automatically remove these actions or change to a different runtime kind.
+4. Once the system's action artifact store does no more contain actions with the runtime kind to be removed, remove the runtime kind from the [runtimes manifest](actions-new.md#the-runtimes-manifest).
+5. TODO - is this correct?: Remove the runtime kind from the list of known kinds in the `ActionExec` object of the [controller API's Swagger definition](../core/controller/src/main/resources/apiv1swagger.json).
```

Comment:
@rabbah Is step 5 required? Is anything else missing?

Reply:
It's important to tell the API caller what kinds of runtimes supporting, so I think it's required to do so.

Reply:
I'm not sure where the Swagger is exactly used. The controller itself offers an endpoint where it publishes the supported runtimes - and this seems to base on the runtimes manifest, not the Swagger definition. See for example: https://us-south.functions.cloud.ibm.com/

Reply:
You're right about this, the `Controller.info()` use runtimes manifest.
So the `ActionExec` in `apiv1swagger.json` would only make sense when executing unit tests.

Reply:
Actually, I wonder whether a fixed list of supported runtime kinds in the Swagger API definition makes sense at all...

* We expect that each adopter will have their own set of supported runtime families and kinds resulting in a customised version of the `runtimes.json` file.
* We have a mechanism in Openwhisk that allows adopters to configure their own `runtimes.json` instead of using the default file: https://github.com/apache/openwhisk/blob/06362ad83608335a01ccbb4ef449ec4c0e70dd99/ansible/group_vars/all#L54-L69.
* In order to obtain a sound Swagger API definition, each adopter has to synchronise [resource file `apiv1swagger.json`](../core/controller/src/main/resources/apiv1swagger.json).
  Maybe I'm missing something - but I see no configuration mechanism to replace the resource file provided by Openwhisk. So replacing this file requires some extra operations...
* At the same time, the controller provides a REST endpoint where it publishes all currently supported runtime families and kinds. This information comes from the configured `runtimes.json` file.
* Wouldn't it make sense to remove the runtime kind `enum` from the Swagger API definition (https://github.com/apache/openwhisk/blob/06362ad83608335a01ccbb4ef449ec4c0e70dd99/core/controller/src/main/resources/apiv1swagger.json#L1890-L1915) and add a hint to the description that there is a REST endpoint where the current list can be obtained?
  Does anybody know whether the `enum` data is used somewhere?

Reply:
Don't get any idea about it.

Remove it and run the tests would make sense.
[Local development](https://github.com/apache/openwhisk/tree/master/ansible/README.md) and [Unit Test](https://github.com/apache/openwhisk/tree/master/tests/README.md)

Reply:
I would support to remove this enum from the swagger documentation and having a pointer to the REST endpoint instead. 
We may think about mentioning two or three samples of stable/longer time supported runtimes, e.g. nodejs:12, nodejs:default, python:3 in this description, just that users get a feeling how the values can look like, but do not specify the whole list.

I have not seen unit tests, that rely on this swagger enum, yet, but in case they exist, I would prefer that we change them to query the REST endpoint, instead of querying the info from the swagger documentation.

Reply:
I agree with @sven-lange-last's  line of thought.  runtimes.json is the source of truth, so we should be using it via the controller's REST API, not an out-of-synch swagger.

Reply:
As discussed on tech exchange, updating the swagger and the implications of that is best done in a separate follow up PR.